types/query/pagination.go

```diff
@@ -0,0 +1,99 @@
+package query
+
+import (
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/store/types"
+)
+
+// maxLimit is the maximum allowed `limit` for queries
+// if the `limit` is not supplied, paginate will use `maxLimit` as default `limit`
+const maxLimit = 100
+
+// Paginate does pagination of all the results in the PrefixStore based on the
+// provided PageRequest. onResult should be used to do actual unmarshaling.
+//
+// Ex:
+//		prefixStore := prefix.NewStore(store, someRequestParam)
+//		var results []Result
+//		pageRes, err := query.Paginate(accountStore, req.Page, func(key []byte, value []byte) error {
+//			var result Result
+//			err := Unmarshal(value, &balance)
+//			results = append(results, result)
+//			...
+//		})
+func Paginate(
+	prefixStore types.KVStore,
+	req *PageRequest,
+	onResult func(key []byte, value []byte) error,
+) (*PageResponse, error) {
+	offset := req.Offset
+	key := req.Key
+	limit := req.Limit
+
+	if offset > 0 && key != nil {
+		return nil, fmt.Errorf("invalid request, either offset or key is expected, got both")
+	}
+
+	if limit == 0 || limit > maxLimit {
+		limit = maxLimit
```

Comment:
I also think we should enable could total if `limit == 0`

Reply:
>  I think we should have a default limit, but not a max limit.

I agree. Maybe even the max limit should be supplied by the caller?

Reply:
Okay, makes sesnse. We are going to remove `maxLimit` and use `defaultLimit`